precompiles/randomness/src/lib.rs

```diff
@@ -0,0 +1,462 @@
+		let status =
+			if let Some(RequestState { request, .. }) = Pallet::<Runtime>::requests(request_id) {
+				if request.is_expired() {
+					RequestStatus::Expired
+				} else if request.can_be_fulfilled() {
+					RequestStatus::Ready
+				} else {
+					RequestStatus::Pending
+				}
+			} else {
+				RequestStatus::DoesNotExist
+			};
+		Ok(succeed(EvmDataWriter::new().write(status).build()))
+	}
+	fn get_request(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let request_id: u64 = handle
+			.read_input()?
+			.read::<U256>()?
+			.try_into()
+			.map_err(|_| error("Input RequestId overflows u64 type set in Pallet"))?;
+		// record cost of 2 DB reads
+		handle.record_cost(RuntimeHelper::<Runtime>::db_read_gas_cost() * 2)?;
+		let RequestState { request, .. } =
+			Pallet::<Runtime>::requests(request_id).ok_or(error("Request Does Not Exist"))?;
+		let status = if request.is_expired() {
+			RequestStatus::Expired
+		} else if request.can_be_fulfilled() {
+			RequestStatus::Ready
+		} else {
+			RequestStatus::Pending
+		};
+		let (
+			randomness_source,
+			fulfillment_block,
+			fulfillment_epoch,
+			expiration_block,
+			expiration_epoch,
+			request_status,
+		) = match request.info {
+			RequestInfo::BabeEpoch(epoch_due, epoch_expired) => (
+				RandomnessSource::RelayBabeEpoch,
+				0u64,
+				epoch_due,
+				0u64,
+				epoch_expired,
+				status,
+			),
+			RequestInfo::Local(block_due, block_expired) => (
+				RandomnessSource::LocalVRF,
+				block_due
+					.try_into()
+					.map_err(|_| revert("block number overflowed u32"))?,
+				0u64,
+				block_expired
+					.try_into()
+					.map_err(|_| revert("block number overflowed u32"))?,
+				0u64,
+				status,
+			),
+		};
+		let (refund_address, contract_address, fee): (Address, Address, U256) = (
+			request.refund_address.into(),
+			request.contract_address.into(),
+			request.fee.into(),
+		);
+		let writer = EvmDataWriter::new()
+			.write(request_id)
+			.write(refund_address)
+			.write(contract_address)
+			.write(fee)
+			.write(request.gas_limit)
+			.write(request.salt)
+			.write(request.num_words)
+			.write(randomness_source)
+			.write(fulfillment_block)
+			.write(fulfillment_epoch)
+			.write(expiration_block)
+			.write(expiration_epoch)
+			.write(request_status)
+			.build();
+		Ok(succeed(writer))
+	}
+	/// Make request for babe randomness one epoch ago
+	fn request_babe_randomness(handle: &mut impl PrecompileHandle) -> EvmResult<PrecompileOutput> {
+		let mut input = handle.read_input()?;
+		let contract_address = handle.context().caller;
+		let refund_address = input.read::<Address>()?.0;
+		let fee: BalanceOf<Runtime> = input
+			.read::<U256>()?
+			.try_into()
+			.map_err(|_| revert("amount is too large for provided balance type"))?;
+		let gas_limit = input.read::<u64>()?;
+		let salt = input.read::<H256>()?;
+		let num_words = input.read::<u8>()?;
+		handle.record_cost(REQUEST_RANDOMNESS_ESTIMATED_COST)?;
+		let two_epochs_later =
```

Comment:
We probably need to add one additional db read for the BabeDataGetter below

Reply:
It might make more sense (in the future) to benchmark these precompile fns directly.

One thing that is "new" with this PR is a shift in logic out of the pallet and into the procempiles. Until now, our precompiles are generally just small wrappers that bridge our pallets and Ethereum smart contracts.

This has both security and performance implications. When our precompiles are minimal wrappers, we can focus our security and performance (both optimization and measurement) entirely at the pallet-level, but that's not true here.

Reply:
Added the db read. Will keep this unresolved to keep comment above viewable.

> It might make more sense (in the future) to benchmark these precompile fns directly.

Interesting idea